Source/ASDisplayTree.mm

```diff
@@ -0,0 +1,278 @@
+//
+//  ASDisplayTree.m
+//  AsyncDisplayKit
+//
+//  Created by Adlai on 5/28/18.
+//  Copyright © 2018 Pinterest. All rights reserved.
+//
+
+#import "ASDisplayTree.h"
+
+#import <AsyncDisplayKit/ASDisplayNodeInternal.h>
+#import <AsyncDisplayKit/ASThread.h>
+
+using namespace ASDN;
+
+/*
+ * Assert we're locked. ASThread.h has some of the same logic, but it's not on
+ * by default during debugging for performance reasons. Since this component is much
+ * higher level, and also since it requires the client to lock it for enumerating,
+ * this assertion is at the default level.
+ */
+#define ASDisplayTreeAssertLocked() ASDisplayNodeAssert(_lockedThread == pthread_self(), nil)
+#define ASDisplayTreeAssertNotLocked() ASDisplayNodeAssert(_lockedThread != pthread_self(), nil)
+
+@interface ASDisplayTree () <NSFastEnumeration>
+@end
+
+@implementation ASDisplayTree {
+  ASDN::RecursiveMutex _lock;
+#if ASDISPLAYNODE_ASSERTIONS_ENABLED
+  pthread_key_t _lockSpecific;
+#endif
+  
+  ASPrimitiveTraitCollection _traitCollection;
+}
+
+#pragma mark - Querying
+
+- (NSUInteger)l_indexOf:(ASDisplayNode *)node
+{
+  NSParameterAssert(node);
+  ASDisplayTreeAssertLocked();
+  auto parent = [self supernodeOf:node];
```

Comment:
My vote is that auto in this context is ultimately harder to read, especially when mixing object and scalar types. As a reader, it's necessary to do the type inference on what the method is going to return, rather than to see it clearly in the code.

It's fine if you'd prefer to land it this way, just something to think about as we consider how the community wants to promote increased usage of auto (clearly it is valuable in many parts of our codebase, especially with all the Objective-C++).